hazelcast/src/main/java/com/hazelcast/instance/DefaultAddressPicker.java

```diff
@@ -337,28 +344,54 @@ private static AddressDefinition pickLoopbackAddress(String host, int defaultPor
     AddressDefinition pickMatchingAddress(Collection<InterfaceDefinition> interfaces) throws SocketException {
         Enumeration<NetworkInterface> networkInterfaces = NetworkInterface.getNetworkInterfaces();
         boolean preferIPv4Stack = preferIPv4Stack();
-        boolean matchInterfaceDefinition = CollectionUtil.isNotEmpty(interfaces);
+        boolean preferIPv6Addresses = preferIPv6Addresses();
+        AddressDefinition matchingAddress = null;
+
+        // There are 3 possible value pairs for preferIPv4Stack & preferIPv6Addresses:
+        // - preferIPv4Stack=true, preferIPv6Addresses=false: Only an IPv4 address will be picked.
+        // - preferIPv4Stack=false, preferIPv6Addresses=false: Either an IPv4 or IPv6 address may be picked, no preference.
+        // - preferIPv4Stack=false, preferIPv6Addresses=true: Either an IPv4 or IPv6 address may be picked
```

Comment:
This part feels a bit counter intuitive. Imho preferIPv4Stack and preferIPv6Stack should be mutually exclusive and always one should be selected. 

Using 2 boolean properties for a binary behavior (either ipv4 or ipv6) is the root cause.

Reply:
They are JDK system properties, we have not introduced them. But still they are not mutually exclusive. They are a bit different, as their names imply too: `preferIPv4Stack` vs `preferIPv6Addresses`. Please see https://docs.oracle.com/javase/8/docs/api/java/net/doc-files/net-properties.html

Reply:
- `java.net.preferIPv4Stack` (default: false)
If IPv6 is available on the operating system the underlying native socket will be, by default, an IPv6 socket which lets applications connect to, and accept connections from, both IPv4 and IPv6 hosts. However, in the case an application would rather use IPv4 only sockets, then this property can be set to true. The implication is that it will not be possible for the application to communicate with IPv6 only hosts.

- `java.net.preferIPv6Addresses` (default: false)
When dealing with a host which has both IPv4 and IPv6 addresses, and if IPv6 is available on the operating system, the default behavior is to prefer using IPv4 addresses over IPv6 ones. This is to ensure backward compatibility, for example applications that depend on the representation of an IPv4 address (e.g. 192.168.1.1). This property can be set to true to change that preference and use IPv6 addresses over IPv4 ones where possible.